# FEATURE - Setup Merger

blackironj opened the pull request:
### 추가사항

### admin plugin
- Admin plugin 에서 `setup` 명령을 처리
  - storage에 user의 sk와 cert 정보가 없을 경우:
  - Setup 명령을 받으면 설정한 `setup-port`를 이용해 임시 setup server를
생성합니다. (Sync server이며 User로 부터 정보를 받을 때 까지 block
된 상태를 유지합니다. )  User로 부터 `MSG_MERGER_SETUP` 을 받아 parsing 후 처리

### TODO
- Start / Stop 명령 처리
- storage에서 user key 가 저장되어 있는지 확인 및 받아오기. 
- Setup 여부에 따라 merger 동작 방식 수정 
  - ex) observer mode ,  tracker 접속 유무 등.

blackironj commented:
- [channel.hpp](https://github.com/gruut/gruut/pull/115/files#diff-7a731046158467f2c60b21bd80f99dbc)
에서 `publishAndGetRes` 는 json 객체를 return 받도록 하였습니다.  이유는 좀 더 리턴값을 좀더 폭넓게 사용할 수 있을 거라 생각하였습니다.

- template으로 좀더 잘 만들 수 있을거 같은데 어떻게 해야할지 감이 안오네요.  수정 할 사항이나 고칠 부분 있으면 말씀해주세요.

morishjs commented on lib/appbase/include/channel.hpp:
```diff
@@ -52,22 +54,42 @@ class Channel : public AbstractChannel {
     }
   }
 
+  nlohmann::json publishAndGetRes(Data &data) {
```
위의 publish의 경우에 proactor를 하나로 두어서(one thread) 나름 채널이 thread-safe 하다는 것을 보장하는데 publishAndGetRes는 race condition이 발생할 것 같습니다. 밑에 채널 `net_control`를 다수의 쓰레드가 접근해서 각각 signal을 보낸다고 생각해보면 쉬울 것 같습니다.

blackironj replied:
문제가 있을거같네요. 일단 net- plugin 을 거치지 않고, admin-plugin에서만 처리하는걸로 수정했을때, 위와 같이 `channel.hpp`를 수정 할 필요가 없을 거 같습니다.

morishjs reviewed (commented):
Q1. 데이터 플로우가 `AdminPlugin` -> (Setup Message 전달) -> `NetPlugin 에서 처리` -> (channel로 AdminPlugin 에게 전달) 하는 방식인데 AdminPlugin에서 모든 걸 처리하면 될 것 같은데 왜 NetPlugin 으로 데이터를 전달해야 하죠?
Q2. 이거 사용 시나리오가 어떻게 되나요? 제가 콘솔로 
`grunt-console setup` 이라는 명령어를 호출할때 명령어 파라미터로 뭘 넘겨줘야 하죠? 코드만 봤을때는 비밀번호만 넘겨주면 될 것 같은데

blackironj commented:
@park53kr 
1. 생각해보니 굳이 net-plugin을 거칠 필요 없이 admin-plugin 에서만 처리해도 될 거같네요. 이 부분 수정할게요. 
2. 정확히 정해진건 없는데, 제 생각에는 현재 비밀번호 외에 넘겨줄 건 없을거라 생각합니다.

- storage에서 key 저장 / 가져오기 가능할 때, key를 가져 오도록 하는 부분은 channel을 통하거나 직접 가져오도록 하여, 추후 관련 내용에서 작성하도록 할게요.

blackironj commented:
수정 후 리뷰 다시 요청드릴게요.

Commit:
# MOVE - MessageParser class

- MessageParser class 이동 ( rpc_service.cpp -> message_parser.hpp )
 - 다른 곳에서도 사용하기 위함.

Commit:
# ADD - Handle setup command ( Admin - plugin )

- setup 명령을 처리
 - storage에 user의 sk와 cert 정보가 없다면, 임시 setup server를 열어
user로 부터 key 정보를 받을 수 있도록 함.

morishjs commented on src/plugins/admin_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -9,9 +11,85 @@ namespace admin_plugin {
+    setup_server->Shutdown();
+  setup_server.reset();
+
+  if (user_key_info.get().has_value()) {
```
코드를 안돌려봐서 모르겠지만, 45번줄에서 setup_server 를 reset하면 promise - future 간 채널이 닫히면서 user_key_info.get()이 정상동작할지 의문이 드네요.

blackironj replied:
`setup_service` 는 `setup_server`와는 다른 별도의 객체입니다. 위에서 user_key_info.wait()에서 future에 대한 값을 받을 때까지 기다리므로 `setup_server`가 닫힌다고 해서 문제 될거 같진 않습니다만

코드
```
  if (setup_server != nullptr)
    setup_server->Shutdown();
  setup_server.reset();
```
아래쪽으로 내리도록 할게요.

아래쪽으로 내리려고했는데,  조금 문제가 있네요 이쪽 코드부분 조금 수정해서 다시 리뷰 부탁드릴게요.

morishjs commented on src/plugins/admin_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -9,9 +11,85 @@ namespace admin_plugin {
+  string setup_serv_addr("0.0.0.0:" + port);
+  SetupService setup_service;
+
+  ServerBuilder setup_serv_builder;
```
새로운 Server를 임시로 생성해야 하는지 PR에서 안적혀있어서 이해가 안되는데 코멘트 부탁드립니다.

Commit:
# EDIT - Run setup server

 #### 문제점
- 기존 코드에서 SetupService에서 promise에 값을 세팅하였을 시점에서
Status::OK를 하지 않음
- Status::OK를 하지 않은시점에서 setup_server를 shutdown 할 경우 문제
발생할 수 있음.

 ##### 해결
- Setup 과정이 모두끝난 후 shutdown과 reset 진행

blackironj commented:
@park53kr 
shutdown 과 reset에서 다른 문제가 있을거 같아서 수정하였습니다.
`SetupService::UserService()`에서 promise에 값을 set 하면 바깥에서 wait() 하던부분을 빠져나와서 shutdown과 reset을 하게되는데 이 시점이 `SetupService::UserService()` 에서 return Status::OK를 하기 전 시점이 될 수도 있을 거 같아서 수정하였습니다.

Setup 과정이 전부 끝나고 shutdown 과 reset을 하도록 수정하였습니다.
